statistics/histogram.go

```diff
@@ -43,26 +46,105 @@ type Histogram struct {
+	// Bucket counts are always in increasing order.
+	//
+	// A bucket value is the greatest item value stored in the bucket.
+	//
+	// Repeat is the number of repeats of the bucket value, it can be used to find popular values.
+	Bounds  *chunk.Chunk
+	Counts  []int64
+	Repeats []int64
+
+	// Used for estimating fraction of the interval [lower, upper] that lies within the [lower, value].
+	// For some types like `Int`, we do not build it because we can get them directly from `Bounds`.
+	lowerScalar  []float64
+	upperScalar  []float64
+	commonPfxLen []int // commonPfxLen is the common prefix length of the lower bound and upper bound when the value type is KindString or KindBytes.
 }
 
-// Bucket is an element of histogram.
-//
-// A bucket count is the number of items stored in all previous buckets and the current bucket.
-// bucket numbers are always in increasing order.
-//
-// A bucket value is the greatest item value stored in the bucket.
-//
-// Repeat is the number of repeats of the bucket value, it can be used to find popular values.
-//
-type Bucket struct {
-	Count        int64
-	UpperBound   types.Datum
-	LowerBound   types.Datum
-	Repeats      int64
-	lowerScalar  float64
-	upperScalar  float64
-	commonPfxLen int // when the bucket value type is KindString or KindBytes, commonPfxLen is the common prefix length of the lower bound and upper bound.
+// NewHistogram creates a new histogram.
+func NewHistogram(id, ndv, nullCount int64, version uint64, tp *types.FieldType, bucketSize int) *Histogram {
+	return &Histogram{
+		ID:                id,
+		NDV:               ndv,
+		NullCount:         nullCount,
+		LastUpdateVersion: version,
+		tp:                tp,
+		Bounds:            chunk.NewChunkWithCapacity([]*types.FieldType{tp, tp}, bucketSize),
+		Counts:            make([]int64, 0, bucketSize),
+		Repeats:           make([]int64, 0, bucketSize),
+	}
+}
+
+// GetLower gets the lower bound of bucket `idx`.
+func (hg *Histogram) GetLower(idx int) *types.Datum {
+	d := hg.Bounds.GetRow(idx).GetDatum(0, hg.tp)
+	return &d
+}
+
+// GetUpper gets the upper bound of bucket `idx`.
+func (hg *Histogram) GetUpper(idx int) *types.Datum {
+	d := hg.Bounds.GetRow(idx).GetDatum(1, hg.tp)
+	return &d
+}
+
+// AddBucket adds a bucket into `hg`.
+func (hg *Histogram) AddBucket(lower *types.Datum, upper *types.Datum, count, repeat int64) {
+	hg.Counts = append(hg.Counts, count)
+	hg.Repeats = append(hg.Repeats, repeat)
+	hg.Bounds.AppendDatum(0, lower)
+	hg.Bounds.AppendDatum(1, upper)
+}
+
+func (hg *Histogram) updateLastBucket(upper *types.Datum, count, repeat int64) {
+	len := hg.NumBuckets()
+	lower := hg.GetLower(len - 1)
+	hg.Bounds.TruncateTo(len - 1)
+	hg.Bounds.AppendDatum(0, lower)
+	hg.Bounds.AppendDatum(1, upper)
+	hg.Counts[len-1], hg.Repeats[len-1] = count, repeat
+}
+
+// DecodeTo decodes the histogram bucket values into `tp`.
+func (hg *Histogram) DecodeTo(tp *types.FieldType, timeZone *time.Location) error {
```

Comment:
Maybe ConvertToType is a better name. This function is to convert the old histogram to one with new type.

Reply:
Actually `DecodeTo` is better, because we decode the value into `tp`. There is another function do the `convert`.